### ISSUE TITLE
Highlighting packages

### ISSUE THREAD

**Issue by mrpatpat** [MaibornWolff/codecharta#152]
Original feature request was #119. Depends on #155.

## Description

3D visualization should color packages in grayscale colors. Root should be the darkest one and colors should get brighter with the depth.
The 3D visualization should allow highlighting packages permanently (by right clicking the tree view element or 3d element and then selecting highlight in a context menu).
The colors per node level should be distinct to each other.

## Definition of Done

A task/pull request will not be considered to be complete until all these items can be checked off.

* [ ] **All** requirements mentioned in the issue are implemented
* [ ] Does match the Code of Conduct and the Contribution file 
* [ ] Task has its own **GitHub issue** (something it is solving)
  * [ ] Issue number is **included in the commit messages** for traceability
* [ ] **Update the README.md** with any changes/additions made
* [ ] **Update the CHANGELOG.md** with any changes/additions made
* [ ] **Enough test coverage to ensure that uncovered changes do not break functionality**
* [ ] **All tests pass**    
* [ ] **Descriptive pull request text**, answering:
  + What problem/issue are you fixing?
  + What does this PR implement and how? 
* [ ] **Assign your PR to someone** for a code review
  + This person _will be contacted **first**_ if a bug is introduced into `master`
* [ ] **Manual testing** did not fail

**Comment by ukinimod** [comment]
Personally I don`t like grayscales and on the other hand highlighting of depth probably only makes sense up to some level. But if it helps this would be great. But this is rather independent of the highlighting.

Concerning manual highlighting permanently I disagree with this feature. In all my usecases I almost always know the blocks that I would like to highlight and it would be rather cumbersome to need to click them in the UI compared to provide the cc.json with the necessary information. Highlighting would also never ever change for one cc.json, at least after having adjusted the right level (see #158).

**Comment by mrpatpat** [comment]
As mentioned in #158 I like the idea of adding the data to the cc.json file. I also think that adding a UI Interaction to additionally highlight/deselecting building is a simple and small customization option. If we implent #158 then just a ssimple template adjustment is everything we need.

**Comment by mrpatpat** [comment]
I implemented some functionality for this. 

1. cc.json has an optional property called markingColor which accepts a string with a hex color e.g. 0xff00aa.
2. context menu offers three marking colors. Clicking the option colors the selected folder and its children (we could implement more selectors like "color all of this level", "color parents", "color children", ...)
3. marking colors do not use a gradient but alternate between even and odd. 
4. floor color is a mix of gradient and alteration. This allows visual distinction between levels and overall height
5. context menu also offers options like isolate or hide (isolates or hides node and autofits camera, my impression is that this could be very practical.)
6. buildings are more pale (maybe too pale, but please remember it is a work in progress)
7. reduced overall map height

While developing this I learned some lessons about visualizing:

1. Pale buildings seem less chaotic because they are similiar in color tone and make up visual groups
2. A dark floor separates these groups
3. Using bright colors for highlighting separates floor and buildings even more. At the same time it groups everything in its highlighting zone strongly. 
4. A dark floor lets the highlight colors "pop up" more 
5. Height gradients are nice but only combined with alteration between even and odd. 
6. A flatter map (less height overall) makes margins and highlightings more dominant since they are not hidden behind buildings. It also looks more like a city/map and not like a bunch of pillars. 

Here is a screenshot of the highlights and changes:
![capture](https://user-images.githubusercontent.com/2622069/40879917-5b50dd0a-66a8-11e8-950a-cf50f4442b7c.PNG)

The current state can be tested by checking [this branch](https://github.com/MaibornWolff/codecharta/compare/feature/grey_scale_packages_treemap_types) out. 

In my opinion this makes the map structure a lot more visible. Additionally there are tasks to implement margins and labels. If we decrease margin between buildings (grouping) and increase it between packages (separation) we should get pretty good results. 

What do you think ?

**Comment by ukinimod** [comment]
@mrpatpat making it possible to define the highlighting via ui and via ccsh sounds nice. Will probably allow both ui centered user and shell centered user a possibility to have their visualization. As user the latter would be my highly prefered solution. If it's only possible to extend it to the interface between analysis and visualization I will reconsider using this feature altogether.

@mrpatpat concerning the implementation.
1. didn't try that yet. 
2. Context menu in the map worked for me, but it was very cumbersome to click the right packages, I had to use seperating via tree view and then click... Context menu in the tree view didn't work.

Actually the branch contains very much changes that are quite hard to evaluate separately. I like the direction of the visualization, but I think it goes a bit to far. Hotspot identification is difficult with these very pale colors (maybe there is a sweet spot between the old implementation) and the overall height.

Three colors is definetly too few for my usecases. Especially as neighbored folder that are of the same color can not be distinguished in the visualization.

Thx, for setting the margin for a more reasonable default.

**Comment by mrpatpat** [comment]
@ukinimod Thank you for your help.

I reverted the building colors and added more possible highlighting colors. The context menu is now also available in the tree view. How many colors could be a good amount ? Maybe we should define a color palette and additionally let the user select a color from a color picker ? 

![capture](https://user-images.githubusercontent.com/2622069/40943353-0707e1bc-6851-11e8-95d0-b33550dc8985.PNG)

As for distinctive neighbours with the same color: #151 should add another distinction layer which should make it possible. 

Regarding the code: This propably could have been split into more issues but I figured that many of these changes are necessary or easy to fix/implement while implementing this issue. I will pay more attention to splitting in the future.
